### ISSUE TITLE
net/http: ServeMux redirect loses query info

### ISSUE THREAD

**Issue by sdwarwick** [golang/go#17841]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?

go version go1.7.1 windows/amd64

### What operating system and processor architecture are you using (`go env`)?

go version go1.7.1 windows/amd64

### What did you do?

this error cannot be seen on play as it requires http.

```
package main

import (
	"fmt"
	"net/http"
)

func main() {

	http.HandleFunc("/testOne", testOneHandler)  // no slash at end or you lose the query string
	http.HandleFunc("/testTwo/", testTwoHandler) // no slash at end or you lose the query string

	http.ListenAndServe(":8008", nil)

}

func testOneHandler(w http.ResponseWriter, r *http.Request) {

	textOut := r.URL.Query()
	fmt.Fprintln(w, textOut)

	return
}

func testTwoHandler(w http.ResponseWriter, r *http.Request) {

	textOut := r.URL.Query()
	fmt.Fprintln(w, textOut)
	return
}
```

### What did you expect to see?

This version works:
input URL:  http://localhost:8008/testOne?this=that
result:  URL stays the same, and the output is:  map[this:[that]]

This doesn't:

input URL:  http://localhost:8008/testTwo?this=that

result:  URL is redirected to http://localhost:8008/testTwo/
    there is no output.

conclusion: 
    query strings are lost when the input URL and the http router have a "/" at the end of the selector.

   this error was initially reported in #5382,

I hope that this is sufficient information.

**Comment by bradfitz** [comment]
In summary:

```
bradfitz@gdev:~$ curl 'http://localhost:8080/testOne?foo=bar' 
/testOne?foo=bar
bradfitz@gdev:~$ curl 'http://localhost:8080/testOne/?foo=bar'
404 page not found
bradfitz@gdev:~$ curl 'http://localhost:8080/testTwo?foo=bar'
<a href="/testTwo/">Moved Permanently</a>.
 
bradfitz@gdev:~$ curl 'http://localhost:8080/testTwo/?foo=bar'
/testTwo/?foo=bar
```

I assume your question is about my third case? (`http://localhost:8080/testTwo?foo=bar`)

But you said `http://localhost:8080/testTwo/?foo=bar` in your bug report, which I assume is a mistake.

What is your actual bug report?

**Comment by sdwarwick** [comment]
correct, I edited the issue to reflect the typo.  The error is in case #3 where a re-direct occurs and doesn't forward the query string.  Thanks for looking at this.

**Comment by bradfitz** [comment]
Not a regression from Go 1.7, so targetting Go 1.9.

**Comment by odeke-em** [comment]
And here is a playground runnable repro https://play.golang.org/p/t0sLqU5tIS for anyone who wants to test or fix the bug.

**Comment by odeke-em** [comment]
I've mailed https://go-review.googlesource.com/43779.

**Comment by odeke-em** [comment]
I was just whetting my appetite for Kubernetes code and testing it out, then noticed that this bug fix breaks their two of their tests(which seemed to rely on the old buggy behavior)
<img width="1174" alt="screen shot 2017-05-23 at 4 14 28 am" src="https://cloud.githubusercontent.com/assets/4898263/26349857/77ff56dc-3f6e-11e7-8100-f0e2665ee6cf.png">
with the test failing when they encounter:
```
server_test.go:1327: 7: response location: expected http://localhost:12345/exec, got http://localhost:12345/exec?ignore=1&command=ls&command=-a&output=1
```
Just a note as am also going to file a bug on their project for them too, to keep vigilant when they update to Go1.9.

**Comment by bradfitz** [comment]
@odeke-em, nice find! Yeah, if you could send them a PR to make their test work with either Go 1.8 or Go 1.9, that'd save everybody a lot of hassle later. Thanks!

**Comment by odeke-em** [comment]
Aye aye, done sent in https://github.com/kubernetes/kubernetes/pull/46306.

**Comment by odeke-em** [comment]
From offline discussions, the closing CL broke the builds at tip and didn't account for the case where Redirect was invoked directly with a full URL that itself had its own query string. Reopening this issue.

**Comment by odeke-em** [comment]
From offline discussions with @bradfitz, the closing CL broke the builds at tip and didn't account for the case where Redirect was invoked directly with a full URL that itself had its own query string. Reopening this issue.

**Comment by bradfitz** [comment]
We should only change `ServeMux`, not `func Redirect`.

**Comment by namusyaka** [comment]
@odeke-em Could you take a look at the CL?

**Comment by odeke-em** [comment]
Thank you for the ping @namusyaka, and thank you for the CL, I am going to take a look at it in a few.

**Comment by namusyaka** [comment]
@sdwarwick The issue has been fixed by [this commit](ab40107 ). Please let me know if you still have the issue.
